llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -478,51 +478,87 @@ Expected<uint64_t> materializeAbbrevFromTagImpl(MCCASReader &Reader,
   return Size + *MaybePaddingSize;
 }
 
+static Error materializeDebugInfoUnopt(MCCASReader &Reader,
+                                       ArrayRef<cas::ObjectRef> Refs,
+                                       SmallVectorImpl<char> &SectionContents) {
+
+  for (auto Ref : Refs) {
+    auto Node = Reader.getObjectProxy(Ref);
+    if (!Node)
+      return Node.takeError();
+    if (auto F = DebugInfoUnoptRef::Cast(*Node)) {
+      assert(Refs.size() <= 2 &&
+             "If a DebugInfoUnoptRef is seen, there should be no more than 2 "
+             "CAS objects under the DebugInfoSectionRef!");
+      auto Data = F->getData();
+      SectionContents.append(Data.begin(), Data.end());
+      continue;
+    }
+    if (auto F = PaddingRef::Cast(*Node)) {
+      raw_svector_ostream OS(SectionContents);
+      auto Size = F->materialize(OS);
+      if (!Size)
+        return Size.takeError();
+      continue;
+    }
+    llvm_unreachable("Incorrect CAS Object in SectionContents");
+  }
+  return Error::success();
+}
+
 static Expected<uint64_t>
 materializeDebugInfoFromTagImpl(MCCASReader &Reader,
                                 DebugInfoSectionRef SectionRef) {
   SmallVector<cas::ObjectRef> Refs = loadReferences(SectionRef);
-  auto MaybeTopRefs = findRefs<DIETopLevelRef>(Reader, Refs);
   SmallVector<char, 0> SectionContents;
   raw_svector_ostream SectionStream(SectionContents);
+  auto Node = Reader.getObjectProxy(Refs[0]);
+  if (!Node)
+    return Node.takeError();
+  if (auto UnoptRef = DebugInfoUnoptRef::Cast(*Node)) {
+    if (Error E = materializeDebugInfoUnopt(Reader, Refs, SectionContents))
+      return std::move(E);
+  } else {
+    auto MaybeTopRefs = findRefs<DIETopLevelRef>(Reader, Refs);
```

Comment:
could you factor out all this `else` code into a helper function `materializeDebugInfoEfficient` or something like that?